### PR TITLE
add wasm support for FFI

### PR DESCRIPTION
**Pull request by MikeHolman**
Adds ability to call import functions in wasm

[Review on Reviewable](https://reviewable.io/reviews/microsoft/chakracore/336)

### CODE REVIEW

File: lib/WasmReader/WasmByteCodeGenerator.h [hunk 1]
```diff
@@ -54,6 +54,7 @@ namespace Wasm
         }
         Js::FunctionBody * body;
         WasmFunctionInfo * wasmInfo;
+        bool imported;
```
Comment (MikeHolman) [inline lib/WasmReader/WasmByteCodeGenerator.h]: this is redundant, didn't realize when I added that it was already in WasmFunctionInfo. Will remove this and only use one in WasmFunctionInfo

File: lib/Runtime/Base/ScriptContext.cpp [hunk 2]
```diff
@@ -1861,15 +1861,47 @@ namespace Js
...
-                localModuleFunctions[i] = funcObj;
+                if (functionArray[i]->wasmInfo->Imported())
+                {
+                    PropertyRecord const * propertyRecord = nullptr;
+                    LPCUTF8 name = functionArray[i]->wasmInfo->GetName();
+
+                    utf8::DecodeOptions decodeOptions = utf8::doAllowInvalidWCHARs;
+
+                    UINT utf16Len = utf8::ByteIndexIntoCharacterIndex(name, strlen((const char*)name), decodeOptions);
+                    LPCWSTR contents = (LPCWSTR)HeapAlloc(GetProcessHeap(), 0, (utf16Len + 1) * sizeof(WCHAR));
```
Comment (kivancmuslu) [inline lib/Runtime/Base/ScriptContext.cpp]: This is a test comment, please ignore.

Comment (kivancmuslu) [inline lib/Runtime/Base/ScriptContext.cpp]: This is a test comment, please ignore.

Comment (abchatra) [inline lib/Runtime/Base/ScriptContext.cpp]: Who frees HeapAlloc?

**Comment by abchatra** [comment]
:+1:

File: lib/Runtime/Base/ScriptContext.cpp (hunk 2 above)
Comment (MikeHolman) [inline lib/Runtime/Base/ScriptContext.cpp]: no one, apparently. I'll add code to free

File: lib/WasmReader/SExprParser.cpp [hunk 3]
```diff
@@ -166,7 +168,7 @@ SExprParser::ReadExprCore(SExprTokenType tok)
     case wtkBLOCK:
         return ParseBlock();
     case wtkCALL:
-        return ParseCall();
+        return ParseCall();;
```
Comment (JacksonTian) [inline lib/WasmReader/SExprParser.cpp]: extra comma.

**Pull request merged**